### PR TITLE
PBL and Convection and Microphysics update for HR2

### PR DESCRIPTION
**Pull request by Qingfu-Liu**
This PR is created for HR2 with the updated PBL & convection codes (code from Jongil Han and Ruiyu Sun). I have resolved the conflicts of the codes with the current repository. The overall features for the updates:

1) to reduce the negative hurricane intensity biases, a parameterization for environmental wind shear effect is included in the GFS cumulus convection and TKE-EDMF PBL schemes. In addition, the entrainment rates in the GFS cumulus convection and TKE-EDMF PBL schemes are enhanced proportional to the sub-cloud or PBL mean TKE (turbulent kinetic energy) when TKE is larger than a threshold value.

2) A detailed description and test results with the NCEP GFS for the impacts of the new parameterization of environmental wind shear effect and TKE-dependent entrainment enhancement on hurricane intensity and track forecasts are given [here](https://docs.google.com/presentation/d/1ulZ2KPRRjWDCMjA8BvfXsAK1EBYpYVE8/edit?slide=id.p1#slide=id.p1).

3) To enhance the underestimated CAPE forecasts in the GFS, the entrainment rate in the TKE-EDMF PBL scheme is increased as a function of vegetation fraction and surface roughness length. A detailed description and test results with the NCEP GFS for this update are given [here](https://docs.google.com/presentation/d/1WQmkW2QC5lfGQ1mMhqXfecCLGUIoQj97/edit#slide=id.p1).

### CODE REVIEW

**Comment by Qingfu-Liu** [comment]
I have compiled the code with two suites (FV3_GFS_v16, FV3_GFS_v17_p8) and successfully run two tests (control_c384, control_p8).

File: physics/satmedmfvdifq.F [hunk 1]
```diff
@@ -1055,79 +1111,51 @@ subroutine satmedmfvdifq_run(im,km,ntrac,ntcw,ntrw,               &
...
-                tem3 = tem3+(v1(i,1)/dz)**2
-                tem3 = cs0*sqrt(tem3)*sqrt(tke(i,1))
+                tem = tsea(i)*(1.+fv*max(q1(i,1,1),qmin))
+                tem1 = 2.*gotvx(i,n)*(tem-thvx(i,k))
+!               e2(i,n-1) = e2(i,n) + tem1 * dz
+!
+!               tem2 = (u1(i,1)/dz)**2
+!               tem2 = tem2+(v1(i,1)/dz)**2
+!               tem2 = cs0*sqrt(e2(i,n))*sqrt(tem2)
+!
```
Comment (lisa-bengtsson) [inline physics/satmedmfvdifq.F]: You may want to consider removing the lines that are commented out.

**Review by lisa-bengtsson — APPROVED** [review]
Looks good, the reduction of the dry bias is a welcomed contribution.

**Comment by Qingfu-Liu** [comment]
***@***.***, thanks, just removed the comment lines.*

**Review by JongilHan66 — APPROVED** [review]

**Review by JiliDong-NOAA — APPROVED** [review]

File: physics/samfdeepcnv.f [hunk 2]
```diff
@@ -975,7 +979,29 @@ subroutine samfdeepcnv_run (im,km,first_time_step,restart,        &
                endif
             enddo
         endif
-
+!
+!  if tkemean>tkcrt, tem=1+tkemean/tkcrt, clamt=tem*clam
```
Comment (grantfirl) [inline physics/samfdeepcnv.f]: Consider removing these commented lines for cleaner code.

File: physics/samfshalcnv.f [hunk 3]
```diff
@@ -893,6 +895,18 @@ subroutine samfshalcnv_run(im,km,itc,ntc,cliq,cp,cvap,            &
              endif
           endif
         enddo
+!
+!  if tkemean>tkcrt, clamt=(1+tkemean/tkcrt)*clam
```
Comment (grantfirl) [inline physics/samfshalcnv.f]: Consider removing the commented code.

File: physics/satmedmfvdifq.F [hunk 4]
```diff
@@ -1336,7 +1360,8 @@ subroutine satmedmfvdifq_run(im,km,ntrac,ntcw,ntrw,               &
             ptem2 = ptem2 + ptem
 !
 !           tem2 = stress(i)*spd1(i)/zl(i,1)
-            tem2 = stress(i)*ustar(i)*phim(i)/(vk*zl(i,1))
+!           tem2 = stress(i)*ustar(i)*phim(i)/(vk*zl(i,1))
```
Comment (grantfirl) [inline physics/satmedmfvdifq.F]: Consider removing commented code. Old versions will always be saved in git history.

**Review by grantfirl — APPROVED** [review]
Looks good to me.

**Comment by grantfirl** [comment]
@Qingfu-Liu Are you going to open PRs into fv3atm and ufs-weather-model for this?

**Comment by Qingfu-Liu** [comment]
@Grant Firl - NOAA Affiliate ***@***.***> , Thanks for your
suggestions, I will remove the comment lines. I will open PRs for fv3atm
and ufs-weather-model.

**Comment by Qingfu-Liu** [comment]
I just received an e-mail from microphysics developer, he want to make some small changes for HR2, I may combine the code changes to this PR if there is only a small code changes

**Comment by Qingfu-Liu** [comment]
Ruiyu Sun made small changes to  "radiation_clouds.f", and I added him to review the new commit. There are no changes to the existing PR#65

**Review by RuiyuSun — APPROVED** [review]

File: physics/satmedmfvdifq.F [hunk 5]
```diff
@@ -1664,8 +1687,7 @@ subroutine satmedmfvdifq_run(im,km,ntrac,ntcw,ntrw,               &
           endif
           if((pcnvflg(i) .or. scuflg(i)) .and.
      &       (k >= kbx .and. k <= kmx)) then
-            dz = zi(i,k+1) - zi(i,k)
-            tem = f1(i,k) * dz
+            tem = f1(i,k) * del(i,k) / grav
```
Comment (dustinswales) [inline physics/satmedmfvdifq.F]: It looks like there is a [1/g local variable declared in this routine](https://github.com/ufs-community/ccpp-physics/blob/c606ae9ea04a487fcca2c988605d6096becf80e5/physics/satmedmfvdifq.F#L291). Please replace the division by "grav" with the multiplication by "gravi".

File: physics/satmedmfvdifq.F [hunk 6]
```diff
@@ -1720,8 +1742,7 @@ subroutine satmedmfvdifq_run(im,km,ntrac,ntcw,ntrw,               &
       enddo
       do k = 1,km
         do i = 1,im
-          dz = zi(i,k+1) - zi(i,k)
-          tem = f1(i,k) * dz
+          tem = f1(i,k) * del(i,k) / grav
```
Comment (dustinswales) [inline physics/satmedmfvdifq.F]: ditto

**Review by dustinswales — APPROVED** [review]
Looks good. Nice job. Just a small suggestion on using a constant in one of the modules.

**Comment by Qingfu-Liu** [comment]
Thanks. I will change it to multiplication

**Comment by jkbk2004** [comment]
Tests are done at https://github.com/ufs-community/ufs-weather-model/pull/1731. @grantfirl @dustinswales Can you go ahead to merge this pr?

**Pull request merged**